### PR TITLE
Feat: Add Prometheus Postfix exporter

### PR DESCRIPTION
**Pull request by Nimon77**
# Description

Add a Prometheus exporter for Postfix to get some metrics. Useful for making graphs in Grafana, e.g. queued messages

Disabled by default, can be enabled with `ENABLE_EXPORTER=1`.

There may be a better solution, but this one works perfectly.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary, I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] **I have added information about changes made in this PR to `CHANGELOG.md`**

### CODE REVIEW

File: CHANGELOG.md
Comment (georglauterbach) [inline CHANGELOG.md:36]:
```suggestion
- **Features**
  - Prometheus `postfix_exporter`
    - to get metrics from Postfix, such as queued messages
    - environment variable `ENABLE_PROMETHEUS_POSTFIX_EXPORTER` to start the Postfix exporter (default: 0)
    - tests to see if the exporter starts correctly
```

File: Dockerfile
Comment (georglauterbach) [inline Dockerfile:283]:
```suggestion
# -----------------------------------------------
# --- Prometheus Postfix-Exporter ---------------
# -----------------------------------------------

COPY --from=stage-compile /root/go/bin/postfix_exporter /usr/local/bin/prometheus_postfix_exporter
```

File: README.md
Comment (georglauterbach) [inline README.md:53]:
```suggestion
- [Prometheus Postfix Exporter](https://github.com/kumina/postfix_exporter)
```

File: docs/content/config/environment.md
Comment (georglauterbach) [inline docs/content/config/environment.md:363]:
```suggestion
##### ENABLE_PROMETHEUS_POSTFIX_EXPORTER
```

File: target/scripts/build/compile.sh
Comment (georglauterbach) [inline target/scripts/build/compile.sh:38]:
```suggestion
function _compile_postfix_prometheus_exporter() {
```

File: target/scripts/build/compile.sh
Comment (georglauterbach) [inline target/scripts/build/compile.sh:48]:
```suggestion
_compile_postfix_prometheus_exporter
```

File: target/scripts/start-mailserver.sh
Comment (georglauterbach) [inline target/scripts/start-mailserver.sh:169]:
```suggestion
  [[ ${ENABLE_PROMETHEUS_POSTFIX_EXPORTER} -eq 1 ]] && _register_start_daemon '_start_daemon_postfix_prometheus_exporter'
```

File: target/scripts/startup/daemons-stack.sh
Comment (georglauterbach) [inline target/scripts/startup/daemons-stack.sh:47]:
```suggestion
function _start_daemon_postfix_prometheus_exporter { _default_start_daemon 'prometheus-postfix-exporter' ; }
```

File: target/scripts/startup/variables-stack.sh
Comment (georglauterbach) [inline target/scripts/startup/variables-stack.sh:108]:
```suggestion
VARS[ENABLE_PROMETHEUS_POSTFIX_EXPORTER]="${ENABLE_PROMETHEUS_POSTFIX_EXPORTER:=0}"
```

File: target/supervisor/conf.d/exporter.conf
Comment (georglauterbach) [inline target/supervisor/conf.d/exporter.conf:1]:
```suggestion
[program:prometheus-postfix-exporter]
```

File: target/supervisor/conf.d/exporter.conf
Comment (georglauterbach) [inline target/supervisor/conf.d/exporter.conf:8]:
```suggestion
command=/usr/local/bin/prometheus_postfix_exporter
```

File: test/tests/serial/exporter.bats
Comment (georglauterbach) [inline test/tests/serial/exporter.bats:5]:
```suggestion
BATS_TEST_NAME_PREFIX='[Postfix Prometheus Exporter] '
CONTAINER_NAME='dms-test_postfix_prometheus_exporter'
```

File: test/tests/serial/exporter.bats
Comment (georglauterbach) [inline test/tests/serial/exporter.bats:11]:
```suggestion
--env ENABLE_POSTFIX_PROMETHEUS_EXPORTER=1
```

File: test/tests/serial/exporter.bats
Comment (georglauterbach) [inline test/tests/serial/exporter.bats:16]:
```suggestion
```

File: test/tests/serial/exporter.bats
Comment (georglauterbach) [inline test/tests/serial/exporter.bats:24]:
```suggestion
  assert_success
```

File: test/tests/serial/exporter.bats
Comment (georglauterbach) [inline test/tests/serial/exporter.bats:30]:
```suggestion
  assert_success
  assert_output --partial 'postfix_qmgr_messages_inserted_receipients_sum'
```

**Review by georglauterbach — CHANGES_REQUESTED** [review]
First and foremost: very nice PR, thank you for respecting this project's contribution guidelines!

I would like to hear the opinions of other maintainers about this feature.

I'd also like to know how the build stage is affected in terms of time @Nimon77.

I will go ahead and immediately apply my suggestions. They are concerned with consistent naming mostly.

**Comment by georglauterbach** [comment]
Probably needs some very minor formatting still (I most likely messed with some whitespaces in the suggestions), but LGTM 🚀

**Comment by polarathene** [comment]
## CI impact

> I'd also like to know how the build stage is affected in terms of time

- AMD64 (cached) => [1m44s](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11039990165/job/30666870788?pr=4197)
- ARM64 (no cache + QEMU) =>[18m](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11039990165/job/30666979904?pr=4197)

A bit tricky to gage the AMD64 due to cache varying. I've built it locally to compare.

ARM64 presently builds at [16m](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11044915402/job/30681693675), so 2m increase there.

```console
$ docker build --tag local/dms-edge --no-cache .
[+] Building 164.7s (69/69) FINISHED

$ docker build --tag local/dms-go --no-cache .
[+] Building 186.7s (70/70) FINISHED

$ docker image ls | grep docker-mailserver
mailserver/docker-mailserver            edge                        de3367313730   4 days ago      713MB
mailserver/docker-mailserver            14.0                        653944e0f525   3 months ago    710MB

# This PR branch:
$ docker image ls | grep dms-go
local/dms-go                            latest                      140026c4f59b   3 minutes ago   736MB
```

20s overhead locally, CI may be a bit slower but it seems fine. I imagine the builder stage itself has grown given the deps installed to support building the project. As that doesn't leak into bloating the final stage it's only a concern on cost to our CI cache usage.

23MB added weight is not ideal for the functionality, but if other maintainers would like to accept this PR as a temporary addition until Vector is integrated into DMS, I'm fine with that.

---

> I would like to hear the opinions of other maintainers about this feature.

I do appreciate when contributions includes tests and docs, but I'm partially reluctant to approve given the below insights, users interested in this do not require anything from DMS to merge to utilize it. We already provide the means for supporting such without the PR.

I won't block if another maintainer wants this to be included into DMS, but changes will need to be made.

## `postfix_exporter` maintenance

https://github.com/kumina/postfix_exporter

This project doesn't seem to be maintained?
- Last commit was in Oct 2022, and the last release (_`0.3.0` which this PR is contributing_) is from Oct 2021. There is 11 PRs open and 38 issues.
- [Debian already has a package for this](https://packages.debian.org/bookworm/prometheus-postfix-exporter) (_so we don't need to build it_), but there is an [unofficially maintained fork with improvements for k8s environment](https://github.com/kumina/postfix_exporter/issues/117).

Issue reports like [this one](https://github.com/kumina/postfix_exporter/issues/118) (6GB+ separate `mail.log`) are concerning and not likely to be fixed given the obvious lack of maintenance. If we get reports for such, it's not something we can resolve on our end. That makes me uncomfortable to include it into DMS.

---

Since there is already a package in Debian this can be installed in DMS like any other addition that we typically defer users to until sufficient interest justifies inclusion into DMS releases.

I'm more inclined to defer to docs contribution. The test coverage while appreciated isn't anything specific to DMS, nor should we really need an ENV toggle to support this given users do not have any real complexity to extend DMS with this service.

This Go build was not static, but from the looks of it that shouldn't be a problem to do, thus a single portable binary can be used instead (_which is further reason for our feature addition policy to defer integration_):

```console
# Only dependent upon glibc
$ docker run --rm -it local/dms-go ldd /usr/local/bin/prometheus_postfix_exporter
        linux-vdso.so.1 (0x00007fff48d8d000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe7a98f8000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fe7a9adf000)
```

The [mentioned fork](https://github.com/Hsn723/postfix_exporter) would not need to be built either as they're actually publishing releases (static linked amd64 + arm64):

```console
POSTFIX_EXPORTER_URL="https://github.com/Hsn723/postfix_exporter/releases/download/v0.6.0/postfix_exporter_0.6.0_linux_amd64.tar.gz"
curl -fsSL "${POSTFIX_EXPORTER_URL}" | tar -xz -C /usr/local/bin postfix_exporter
```

That's all you need to add the binary if the debian package from 2021 is not acceptable, other than the supervisor service config.

---

## Feature contribution policy (_user demand, maintenance burden, extending DMS_)

We could accept the above curl approach given the fork is technically maintained (_binary is 10MB_), but unless a maintainer sees this as an important addition to DMS, it should be subject to the same policy we've been asking of any other contributors to take. Was there a related feature request issue/discussion prior to this PR being submitted?

Otherwise we'd encourage users to just add a few lines to `compose.yaml`:

```yaml
services:
  dms:
    image: local/dms:14.0.0
    pull_policy: build
    build:
      dockerfile_inline: |
        # syntax=docker/dockerfile:1
        
        FROM mailserver/docker-mailserver:14.0.0
        RUN <<HEREDOC
          VERSION='0.6.0'
          ARCH='amd64'
          URL="https://github.com/Hsn723/postfix_exporter/releases/download/v${VERSION}/postfix_exporter_${VERSION}_linux_${ARCH}.tar.gz"
          curl -fsSL "${POSTFIX_EXPORTER_URL}" | tar -xz -C /usr/local/bin postfix_exporter
        HEREDOC
    configs:
      - source: dms-service-postfix-exporter
         target: /etc/supervisor/conf.d/postfix-exporter.conf

# This could be a volume mount instead, or included as part of the inline `Dockerfile` above via `COPY`:
# `autostart=true` as this is an explicit user customization, no need for ENV + script enabling:
configs:
  dms-service-postfix-exporter:
    content: |
      [program:prometheus-postfix-exporter]
      startsecs=0
      stopwaitsecs=55
      autostart=true
      autorestart=true
      stdout_logfile=/var/log/supervisor/%(program_name)s.log
      stderr_logfile=/var/log/supervisor/%(program_name)s.log
      command=/usr/local/bin/prometheus_postfix_exporter
```

If not wanting that bundled into your `compose.yaml` there is a feature to extend Compose services or other tooling like `yq` that could merge it on demand. Alternatively just have the supervisor config and `Dockerfile` as separate files. Choice is yours.

 `user-patches.sh` works too if avoiding a custom image (_albeit a tad inefficient_):

```yaml
services:
  dms:
    image: mailserver/docker-mailserver:14.0.0
    # ... and everything else, nothing changes here
```

`user-patches.sh` (_part of your existing DMS config volume_):

```bash
#!/bin/bash

curl -fsSL https://github.com/Hsn723/postfix_exporter/releases/download/v0.6.0/postfix_exporter_0.6.0_linux_amd64.tar.gz \
  | tar -xz -C /usr/local/bin postfix_exporter
```

Then add the supervisor service config via volume mount, or create it in `user-patches.sh`. 3-4 lines + 2 files. DMS only runs `user-patches.sh` for fresh containers now, so the startup delay will only occur with new instances:

https://github.com/docker-mailserver/docker-mailserver/blob/94751e00c9787c5223e39bd0cd5b139a0c1f6255/target/scripts/start-mailserver.sh#L188-L190

Slightly inconvenient? Yes, will you see and interact with these additions regularly? No.

**Review by polarathene — CHANGES_REQUESTED** [review]
> Probably needs some very minor formatting still (I most likely messed with some whitespaces in the suggestions)

I saw two lines with incorrect indentation, so yes.

Until I receive maintainer feedback to my larger comment, I'm blocking merging. If the decision is to include this into DMS I'll provide a review for the author with guidance.

**Comment by georglauterbach** [comment]
If this really can be installed with a Debian package, I would also not go ahead with this PR. Thank you for digging into this @polarathene 👍🏼

**Pull request closed without merging**
